File: Объектно объективное програмиование/4 Классы математических объектов с++/Fraction.cpp
#include "Fraction.h"


namespace xDlars
{
	Fraction::Fraction(int a, int b, int c)
	{
		if (b == 0)
			throw"error";

		A = a; B = b; C = c;
		
		if (B < 0)
		{
			A *= -1;
			B *= -1;
		}
		redication();
	
	}


	//copy
	Fraction::Fraction(const Fraction& fraction):Fraction(fraction.A, fraction.B, fraction.C){}

	Fraction& Fraction::operator=(const Fraction& fraction)
	{
		if (this == &fraction)
			return *this;

		A = fraction.A;
		B = fraction.B;
		C = fraction.C;

		return *this;
	}

	void Fraction::redication()
	{
		if (B == 0)
			throw "error";

		if (A != 0 && B != 0)
		{
			int c = NOD();
			A /= c;
			B /= c;
		}
		
	}


	int Fraction::NOD()//наибольший обший множитель
	{
		int a = abs(A); int b = abs(B);
		while (a != 0 && b != 0)
		{
			if (a < b)
				b -= a;
			if (a > b)
				a -= b;
			if (a == b)
				return a;
		}
		if (a == 0)
			return b;
		else
			return a;
	}


	void Fraction::decomposition()//приведение в правильную дробь
	{
		if (abs(A) > abs(B))
		{
			C = abs(A / B);
			A = A % B;
		}
	}

	void Fraction::reDecompositon()//приведение обратнов в неправильную дробь
	{
		if (C == 0) return;

		if (A < 0)
			A = (C * B - A) * (-1);
		else
			A = C * B + A;
		C = 0;
	}


	//-------------------------------------------------------
	Fraction& Fraction::operator+=(const Fraction& drob)
	{
		Fraction d(drob);
		int c;

		if (B != d.B)
		{
			c = d.B;
			d.B = B * d.B;
			d.A = B * d.A;
			A = c * A;
			B = c * B;
		}

		C += d.C;
		A += d.A;
		redication();
		return *this;
	}

	Fraction& Fraction::operator-=(const Fraction& drob)
	{
		Fraction d(drob);
		int c;

		if (B != d.B)
		{
			c = d.B;
			d.B = B * d.B;
			d.A = B * d.A;
			A = c * A;
			B = c * B;
		}

		C -= d.C;
		A -= d.A;
		redication();
		return *this;
	}

	Fraction& Fraction::operator*=(const Fraction& drob)
	{
		Fraction d(drob);

		d.reDecompositon();
		reDecompositon();

		A *= d.A;
		B *= d.B;
		redication();
		return *this;
	}

	Fraction& Fraction::operator/=(const Fraction& drob)
	{
		Fraction d(drob);

		d.reDecompositon();
		reDecompositon();

		A *= d.B;
		B *= d.A;
		if (B == 0)
			throw"error";
		if (B < 0)
		{
			A *= -1;
			B *= -1;
		}
		redication();
		return *this;
	}


	//-----------------------------------
	Fraction operator+(const Fraction& A, const Fraction& B)//Сложение
	{
		return (Fraction(A) += B);
	}

	Fraction operator-(const Fraction& A, const Fraction& B)//Вычитане
	{
		return (Fraction(A) -= B);
	}

	Fraction operator*(const Fraction& A, const Fraction& B)//Произведение
	{
		return (Fraction(A) *= B);
	}

	Fraction operator/(const Fraction& A, const Fraction& B)//Произведение
	{
		return (Fraction(A) /= B);
	}


	void Fraction::output()
	{
		cout << "----------------------" << endl;
		if (C > 0)
			cout << C;
		cout << "( " << A << "/" << B << " )"<<" = " << A/B;


	}
}

File: Объектно объективное програмиование/8 Классы математических объектов с#/Fraction.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Колоев_439_5_B_LB8
{
    struct Fraction
    {
        private int A;
        private int B;

        public Fraction(int a = 0, int b = 1)
        {
            this.A = a;
            this.B = b;
        }

        //copy
        public Fraction(Fraction fraction) : this(fraction.A, fraction.B) { }



        //сокрашение дроби

        private int NOD()//наибольший обший множитель
        {
            int a = Math.Abs(this.A); int b = Math.Abs(this.B);
            while (a != 0 && b != 0)
            {
                if (a < b)
                    b -= a;
                if (a > b)
                    a -= b;
                if (a == b)
                    return a;
            }
            if (a == 0)
                return b;
            else
                return a;
        }

        public void redication()//сокрашение дроби
        {
            if (this.B == 0)
                throw new ArgumentException(" Fraction : error");

            if (this.A != 0 && this.B != 0)
            {
                int c = NOD();
                this.A /= c;
                this.B /= c;
            }
        }

        //правильная дробь
        public int decomposition()//приведение в правильную дробь
        {
            int C = 0;
            if (Math.Abs(this.A) > Math.Abs(this.B))
            {
                C = Math.Abs(this.A / this.B);
                this.A = this.A % this.B;
            }
            return C;
        }


        //перегрузка операторов

        public static Fraction operator +(Fraction A, Fraction B)//Сложение
        {
            Fraction d = new Fraction();

            if (A.B != B.B)
            {
                d.B = A.B * B.B;
                d.A = A.A * B.B + B.A * A.B;
            }
            else
            {
                d.A = A.A + B.A;
                d.B = A.B;
            }

            d.redication();
            return d;
        }
		public static Fraction operator -(Fraction A, Fraction B)//Вычитане
        {
            Fraction d = new Fraction();

            if (A.B != B.B)
            {
                d.B = A.B * B.B;
                d.A = A.A * B.B - B.A * A.B;
            }
            else
            {
                d.A = A.A - B.A;
                d.B = A.B;
            }

            d.redication();
            return d;
        }
        public static Fraction operator *(Fraction A, Fraction B)//Произведение
        {
            Fraction d = new Fraction();

            if(A.A == 0 || B.A == 0)
            {
                d.A = 0; d.B = 0;
                return d;
            }

            d.A = A.A * B.A;
            d.B = A.B * B.B;

            d.redication();
            return d;
        }
        public static Fraction operator /(Fraction A, Fraction B)//Произведение
        {
            Fraction d = new Fraction();

            if (A.A == 0 || B.A == 0)
            {
                d.A = 0; d.B = 0;
                return d;
            }

            d.A = A.A * B.B;
            d.B = A.B * B.A;

            d.redication();
            return d;
        }


        public static implicit operator Fraction(int x)
        {
            return new Fraction(x);
        }
        public static implicit operator int(Fraction counter)
        {
            if (counter.A == 0 || counter.B == 0)
                return 0;
            return counter.A/counter.B;
        }
        public static implicit operator double(Fraction counter)
        {
            if (counter.A == 0 || counter.B == 0)
                return 0;
            return (double)counter.A / (double)counter.B;
        }


        public override string ToString()
        {
            StringBuilder matrixString = new StringBuilder();
            matrixString.Append(" " + A + "/" + B + " ");
            return matrixString.ToString();
        }
    }
}
